### ISSUE TITLE
[Backend][ES] when to backslash ?

### ISSUE THREAD

**Issue by HarishHary** [Neo23x0/sigma#672]
Hello, sigma team

I am currently facing an issue on how to use backslashes to define a string.
Let's say that I want to match the following string on the logs: `C:\\Program Files\\VMware\\VMware Tools\\vmtoolsd.exe`. How should I write the rule ?

As far as I understood, I follow these sets of rules:
> All values are treated as case-insensitive strings
You can use wildcard characters '*' and '?' in strings
Wildcards can be escaped with \, e.g. \*. If some wildcard after a backslash should be searched, the backslash has to be escaped: \\*.
Regular expressions are case-sensitive by default
You don't have to escape characters except the string quotation marks '

So I need to write the following way:
`Image: 'C:\\Program Files\\VMware\\VMware Tools\\vmtoolsd.exe'`

But this generates the following query string:
`Image: 'C\:\\Program\ Files\\VMware\\VMware\ Tools\\vmtoolsd.exe'`

Backslashes are not backslashed like it should ?

Using es-dsl I got the the right query:
```
"match_phrase": {  
  "ProcessName": "C:\\\\Program Files\\\\VMware\\\\VMware Tools\\\\vmtoolsd.exe"
}
```
Where you can see the 4 backslashes.

Using es-qs I got the following query which in my opinion wrong:
`Image:"C\:\\Program\ Files\\VMware\\VMware\ Tools\\vmtoolsd.exe"`

I should see the following query:
`Image:"C\:\\\\Program\ Files\\\\VMware\\\\VMware\ Tools\\\\vmtoolsd.exe"`
 
So why so many backslashes this is what I discovered:
If we look purely at elastic, we use lucene syntax which means that the backslash needs to be escaped AND we need to escape twice because JSON needs to be escaped.

so if we go back to our example:
* in the logs N backslashes =>
* in sigma yaml => 2N backslashes (for single quote and plain scalar strings)
* in generated elastalert yaml => 2N backslashes (nothing changed wrt backslashes)
* in JSON query 4N backslashes 

or taken in another order, might be easier:
* to query agaisnt N backslashes we need 2N backslashes (lucene query syntax)
* then we need to serialize/store that query somehow
  * in JSON backslash is also the escape char => the backslashes in a string needs to be escaped => 4N backslashes
  * in yaml, depends on the type of string, the one we mostly use is single quote or no quote => backslashes don't need to be escaped => 2N backslashes

TLDR: How many backslashes when we observe multiple backslashes in the logs ?
PS: I am also quoting @neu5ron  since he have done a lot of work on the es backend recently.

Best regards,

**Comment by acalarch** [comment]
What log datasource are you trying to match against? Sysmon / 4688 or something else?

**Comment by HarishHary** [comment]
Well log datasource doesn't really matter in this situation, I think. It's more related to how sigma parse the file ? But tests have been done using the following rule file: https://github.com/Neo23x0/sigma/blob/master/rules/windows/builtin/win_susp_time_modification.yml

Hope it does answer your question.

**Comment by acalarch** [comment]
If you need to match on "\\" then you must write the rule as "\\\\" the reason you are seeing "\\\\" in the dsl is because "\\" is being escaped to "\\\\" for json compatibility. At search once the json escaping is removed it will match on "\" in the log.

**Comment by HarishHary** [comment]
> If you need to match on "\" then you must write the rule as "\\" the reason you are seeing "\\" in the dsl is because "\" is being escaped to "\\" for json compatibility. At search once the json escaping is removed it will match on "" in the log.

Did you mean `"\" in the log` ?

I think you missed the raised issue here... Maybe I didn't explain it really well.
Here how sigma works as far as I understood. You write the string as you want to match on the logs (exception excluded). So if you want to match `C:\Windows`, you write it as `C:\Windows` it will get translated to `C\:\\Windows` => which is the correct behaviour.

Now what is not expected is the following: When you want to match `C:\\Windows` if you write it as is `C:\\Windows` it generate the following query `C\:\\Windows` which is false ! It should get tranlated to `C\:\\\\Windows`.

Do you follow me now ?

**Comment by acalarch** [comment]
I did.

So you want some logic like: If '\\' not followed by a special character except '\' in sigma (i.e. * ?) than treat it as the rule creator meant to match on "\\" not "\"

There will still be instances where the backslash will have to be escaped such as when a wildcard follows a backlash (i.e. to match on C:\windows\*).. a pretty common thing to do. 

I think it makes sense to keep it the way it is.. although there are some current rules for UNC paths that are affected by the assumption that \\  matches \\ and not just \.

**Comment by HarishHary** [comment]
Ok this behavior should then be documented. Anyhow, when I tried to apply the fix given by @neu5ron on this issue: https://github.com/Neo23x0/sigma/issues/662
It somehow fixed this issue as-well, but it may have broken others too... And investigated further or at least tell the rule creator a strict way to write the rules. Either the rule creator handles the backslashes or sigmas handles it for us. Not both or It will lead to misunderstanding and expected misbehaviour.
Again, this needs to be documented !

Thank you for your answer. Don't close this issue yet. We might get more answer from @neu5ron , @Neo23x0 or @thomaspatzke.

**Comment by thomaspatzke** [comment]
I think it's important to document the behavior properly, it's now on my todo list.

Unfortunately the topic is quite more complicated than it appears at a first glance as we have to cover various cases (like `\\*`) and also want to keep the possibility to express the most prevalent case of the single backlash in the plain way.

**Comment by thomaspatzke** [comment]
Documented how to backslash correctly in the [wiki](https://github.com/Neo23x0/sigma/wiki/Rule-Creation-Guide#backslashes).
